### PR TITLE
WIP: dev

### PR DESCRIPTION
**Pull request by IsmailM**

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -19,10 +19,13 @@ elif [ "$KERNEL" = "linux" ]; then
...
 fi
 
-# If there is an argument then there is where GV will installed
 # If there is an argument then there is where GV will installed
 if [ "$0" = 'sh' ]; then
+  # I.e. when piping from curl
   INSTALL_DIR=$PWD/genevalidator
+elif [ "$0" = 'install.sh' ]; then
+  # I.e. when running directly
+  INSTALL_DIR=${HOME}/.taxonkit
```
Comment (yeban) [inline install.sh]: Should '.taxonkit' be '.genevalidator'?

File: lib/genevalidator/validation_alignment.rb [hunk 2]
```diff
@@ -124,32 +124,30 @@ def initialize(prediction, hits)
     # Output:
     # +AlignmentValidationOutput+ object
     def run(n = 10)
-      n = opt[:min_blast_hits] < 10 ? 10 : opt[:min_blast_hits]
+      n = opt[:min_blast_hits] < 10 ? n : opt[:min_blast_hits]
```
Comment (yeban) [inline lib/genevalidator/validation_alignment.rb]: This looks like a potentially breaking change. Previously `n` couldn't be lower than 10 (or higher than 50 - L128). With this change, the value of `n` can be set to lower than 10 by the caller.

File: lib/genevalidator/validation_alignment.rb [hunk 3]
```diff
@@ -222,12 +219,11 @@ def run(n = 10)
     # Output:
     # Array of +String+s, corresponding to the multiple aligned sequences
     # the prediction is the last sequence in the vector
-    def multiple_align_mafft(prediction = @prediction, hits = @hits)
-      fail unless prediction.is_a?(Query) && hits[0].is_a?(Query)
+    def multiple_align_mafft(prediction, hits)
+      raise unless prediction.is_a?(Query) && hits[0].is_a?(Query)
 
-      options = ['--maxiterate', '1000', '--localpair', '--anysymbol',
-                 '--quiet', '--thread', @num_threads.to_s]
```
Comment (yeban) [inline lib/genevalidator/validation_alignment.rb]: Do you have statistics on how this changes the runtime?

**Comment by yeban** [comment]
Hey @IsmailM. I don't quite understand what's happening in 081143aa so I will refrain from merging the PR. There's lot of context there that I don't understand. If you are confident of the changes, I suggest you go ahead with merge.

File: lib/genevalidator/validation_alignment.rb (hunk 2 above)
Comment (IsmailM) [reply lib/genevalidator/validation_alignment.rb]: no I don't think this is a breaking change - the variable n in set to 10 in the line above... (so it still cannot be higher than 10...

(This was a change suggested by rubocop)

File: lib/genevalidator/validation_alignment.rb (hunk 3 above)
Comment (IsmailM) [reply lib/genevalidator/validation_alignment.rb]: Not yet - this PR is Work in Progress (WIP) and so will try and put something together....


However, if this increases the run time, I think this is the right way to do this (otherwise we end up requesting a lot more threads than asked for) - which potentially causes things to break completely

**Comment by IsmailM** [comment]
Hey, this is still work in progress - will spend some time on Monday on this...


About 081143a - most of the previous code in these files is from the original version of GV (I might have made minor changes here and there) and that is extremely hard to follow. I feel I have made it easier to follow by splitting into separate functions - not creating an extra variable when a `.map {}` is sufficient etc. Nonetheless, I am pretty confident that this still works in the same way as before - but the code is a little tidier (not perfect, but better than before imo)...

File: lib/genevalidator/validation_alignment.rb (hunk 3 above)
Comment (yeban) [reply lib/genevalidator/validation_alignment.rb]: This can potentially decrease run time as well (running 40 x 40 = 1600 threads on 40 cores may incur considerable overhead in switching thread states). If the improvement can be quantified, it can be mentioned in the changelog (which is ++ for users). 

There are potentially better ways to do this still, for e.g., running 10 MAFFT processes each using 4 threads (that is, assuming the number of threads was set to 40), or 5 processes with 8 threads each, and so on (any m x n combination).

**Comment by IsmailM** [comment]
Hey,

I think this is now ready.

Since you last saw this, I've now added the resuming feature, which works with the preprocessing steps (e.g. BLAST, raw seqs etc.) and also GV itself - i.e. if GV already has results for half of the sequences, it will skip those, and continue with the remaining results ...

**Comment by yeban** [comment]
That's very cool! I will try to review soon-ish

**Comment by yeban** [comment]
Resume worked like a charm on my small test data. Creating a new directory is practical, but also potentially confusing.
- Would be nice if at least `-f` was respected.
- Help text for `-r` is incomplete: ends like this "... (creates a new output directory but skips".

On a side note, would be nice for "The output directory already exists for this fasta file" message to also allude to `-f` option (apart from suggesting to remove the directory altogether).

**Comment by yeban** [comment]
I think I found an edge case. I ran GV without MAFFT. Then tried the resume feature and got the following error:

```
NoMethodError: undefined method `+' for nil:NilClass
  /data/autoScratch/weekly/btw675/genevalidator/lib/genevalidator/output.rb:170:in `block (2 levels) in map_errors'
  /data/autoScratch/weekly/btw675/genevalidator/lib/genevalidator/output.rb:170:in `each'
  /data/autoScratch/weekly/btw675/genevalidator/lib/genevalidator/output.rb:170:in `block in map_errors'
  /data/autoScratch/weekly/btw675/genevalidator/lib/genevalidator/output.rb:168:in `map'
  /data/autoScratch/weekly/btw675/genevalidator/lib/genevalidator/output.rb:168:in `map_errors'
  /data/autoScratch/weekly/btw675/genevalidator/lib/genevalidator/output.rb:117:in `generate_overview'
  /data/autoScratch/weekly/btw675/genevalidator/lib/genevalidator.rb:100:in `produce_output'
  /data/autoScratch/weekly/btw675/genevalidator/lib/genevalidator.rb:59:in `run'
  bin/genevalidator:357:in `<top (required)>'
```
